core/src/test/scala/org/apache/spark/scheduler/BlacklistTrackerSuite.scala

```diff
@@ -0,0 +1,162 @@
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set("1","2"))
+    assert(tracker.executorBlacklist(scheduler, 2L) === Set())
+
+    tracker.updateFailureExecutors(taskInfo_1_1_hostA, Success)
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set("2"))
+    assert(tracker.nodeBlacklist() === Set())
+
+    tracker.updateFailureExecutors(taskInfo_3_3_hostB, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, 3L) === Set("3"))
+    assert(tracker.nodeBlacklist() === Set())
+
+    tracker.updateFailureExecutors(taskInfo_2_1_hostA, Success)
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set())
+  }
+  
+  test ("strict strategy works as exception") {
+    val conf = new SparkConf().setAppName("test").setMaster("local")
+      .set("spark.ui.enabled","false")
+      .set("spark.scheduler.blacklist.strategy", "strict")
+    sc = new SparkContext(conf)
+    val scheduler = new TaskSchedulerImpl(sc, 1)
+
+    val tracker = new BlacklistTracker(conf)
+    tracker.updateFailureExecutors(taskInfo_1_1_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set("1"))
+    tracker.updateFailureExecutors(taskInfo_1_2_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, 2L) === Set("1"))
+    assert(tracker.nodeBlacklist() === Set("hostA"))
+
+    tracker.updateFailureExecutors(taskInfo_2_1_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, 2L) === Set("1", "2"))
+    assert(tracker.nodeBlacklist() === Set("hostA"))
+
+    tracker.updateFailureExecutors(taskInfo_1_1_hostA, Success)
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set("1","2"))
+
+    tracker.updateFailureExecutors(taskInfo_3_3_hostB, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, 3L) === Set("1","2","3"))
+    assert(tracker.nodeBlacklist() === Set("hostA","hostB"))
+  }
+
+  test ("threshold strategy works as exception") {
+    val conf = new SparkConf().setAppName("test").setMaster("local")
+      .set("spark.ui.enabled","false")
+      .set("spark.scheduler.blacklist.strategy", "threshold")
+      .set("spark.scheduler.blacklist.threshold.maxFailureTaskNumber", "1")
+      .set("spark.scheduler.blacklist.threshold.maxBlackExecutorNumber", "0")
+    sc = new SparkContext(conf)
+    val scheduler = new TaskSchedulerImpl(sc, 1)
+
+    val tracker = new BlacklistTracker(conf)
+    tracker.updateFailureExecutors(taskInfo_1_1_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set())
+    tracker.updateFailureExecutors(taskInfo_1_2_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, 2L) === Set("1"))
+    assert(tracker.nodeBlacklist() === Set("hostA"))
+
+    tracker.updateFailureExecutors(taskInfo_2_1_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, 2L) === Set("1"))
+    assert(tracker.nodeBlacklist() === Set("hostA"))
+
+    tracker.updateFailureExecutors(taskInfo_1_1_hostA, Success)
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set())
```

Comment:
this last check is a little bit unusual.  Is it really possible for this task to get scheduled again on executor 1?  It doesn't seem like there would be any way for this task to get scheduled again on this host, before the blacklist is updated.  I suppose its possible, b/c the dagscheduler is told about the task failure before the blacklist is updated, so it could get rescheduled on the same executor before the blacklist is updated -- is that correct?  But the more "normal" route for an executor to get off the blacklist is just the timeout.

its a good thing to test in any case, but I'd put in a comment explaining how you'd end up in this situation.

You should also add a check that if a *different* task succeeds, the executor stays in the blacklist.